### ISSUE TITLE
Correcting years depending on form name

### ISSUE THREAD

**Issue by dianedetoeuf** [OpenFn/ConSoSci#171]
## Describe the bug
When the battery of a smartphone is too low, the date/time is reset to factory time. This causes issue for our analysis as the start time of a survey is for example 2000 instead of 2017. This can't be corrected in Kobo as it's a metadata so I'd like to update it with Open Fn.

## To Check
Can you check the code I've added to try and correct the year depending on the form name, line 136 of this job: https://www.openfn.org/projects/wcs/jobs/jykm75

### expression.js
Link to the job itself in Github: https://github.com/OpenFn/ConSoSci/blob/master/bns/survey.js

**Comment by aleksa-krolls** [comment]
@ritazagoni tomorrow, can you please triage & hand off to Mamadou?

**Comment by ritazagoni** [comment]
hey @lakhassane could you please check and test https://github.com/OpenFn/ConSoSci/blob/4e2714e36fdad71047ed026d76ff6eac2c1471a8/bns/survey.js#L136-L143

**Comment by aleksa-krolls** [comment]
@ritazagoni Do you have time to try this out yourself on devtools today or tomorrow morning? 

Also, if we're going to hand off to engineers, can you please make sure sample Messages & creds are always provided in any request.

**Comment by aleksa-krolls** [comment]
@elias-ba can you please review the code @dianedetoeuf drafted? She is trying to evaluate the year of the `today` date (or `_submission_time` if that is not available), to then determine how to update the year of `SurveyDate`.

It would be great if you could (1) provide feedback on what she drafted, and (2) make any necessary changes so that `SurveyDate` outputs correctly and the run succeeds? 

See commented lines for what she drafted: 
https://github.com/OpenFn/ConSoSci/blob/a539682320dbfab76a8726e1b832c0e222256ad8/bns/survey.js#L134-L144

Here is a Message state.data to test against:
https://www.openfn.org/projects/wcs/messages/060e71cf-6000-7f2b-8ffd-b520f2e65430
For config, see LP `WCS Programs DB`. 
Adaptor is `mssql`

**Comment by elias-ba** [comment]
Hey @aleksa-krolls could you please give me access to the project ? I can't see it currently in OpenFn Platform. As soon as you give me access I'll be able address this issue. Thanks in advance.

**Comment by ritazagoni** [comment]
@elias-ba added you, let me know if any issues

**Comment by elias-ba** [comment]
Thanks @ritazagoni jumping into it right now.

**Comment by elias-ba** [comment]
Hello @dianedetoeuf nice to e-meet you,

I have reviewed your code and I see that you're checking some conditions to update the `surveyDate` value.
I am assuming that the conditions you've written in your commented code below are valid and exhaustive:
```javascript
    //if (Number() <= 2014 and (formName === 'BNS Makira 2019')) {
    // return 2019
    //else if (Number(date.split('-')[0]) <= 2014 and (formName === 'BNS Makira 2017')) {
    // return 2017
    //else if (Number(date.split('-')[0]) <= 2014 and (formName === 'BNS Lac Télé 2017')) {
    // return 2017 
    //} 
    //return date
```
If I understand them well, you're always taking the year in the `formName` value if the year in the `date` is under 2014 otherwise you're taking the year of the date.

Depending on that understanding, I have corrected a little bit your code (see my code below):
```javascript
    const date = state.data.today || state.data._submission_time;
    const year = Number(date.trim().split('-')[0]);
    const formName = dataValue('formName');
    if (year <= 2014) return Number(formName.trim().split(' ').at(-1));
    return year;
```
Basically, what I am saying in this code is that instead of doing a condition for every `formName`, we can just retrieve the year in the `formName` as it's always ending by the year (I confirmed this with @ritazagoni).

I am not sure if this resolves the problem, but please feel free to tell me if my understanding of your code is correct and we can discuss my proposal from there. The run is successful, but I need you to confirm all this.

Thanks

**Comment by dianedetoeuf** [comment]
Hi Elias, 
Nice to e-meet you too! This sounds right, thank you so much. I'll make sure the Form Names end always with the year. I'll run it and let you know if there are any issue.
